### ISSUE TITLE
Create an official docker image

### ISSUE THREAD

**Issue by FireFart** [rapid7/metasploit-framework#5284]
What do you guys think if we create a default docker image for metasploit containing a preconfigured working installation? I have found several user based containers, but no official one.

What do you think? I would love to implement a metasploit install in docker for a easy way to run it if you think it's worth it.

We would need a new github repo for the Dockerfile and a metasploit organization on https://hub.docker.com/

**Comment by mubix** [comment]
I have done this a few times with different installs (Pro trial, framework and Community). I'd be keen on taking this on, make a clean image for each and push them, with a few tutorials on how to get them going

**Comment by FireFart** [comment]
@mubix so if there is a agreement on creating a docker image, maybe we should start a document meet in irc to discuss the implementation details like use rvm or rbenv and such things

**Comment by UserTaken** [comment]
I created [usertaken/metasploit-framework](https://registry.hub.docker.com/u/usertaken/metasploit-framework) after watching [Metasploit Minute](https://www.patreon.com/mubix) so I would be interested on how an official image would be created.

**Comment by timwr** [comment]
Nice one @usertaken. I've played with  github.com/infoslack/docker-metasploit but yours looks neater as it's a single container with both metasploit and postgresql.

**Comment by UserTaken** [comment]
I included PostgreSQL for convenience but alternatively something like 

```
containers:
  loot:
      image: busybox
      run:
          volume: ["/opt/metasploit-framework/.msf4", "/var/www", "/tmp"]
          cmd: ["watch", "-n", "120", "-t", "ls", "-pArcelhR", "--color=always", "/opt/metasploit-framework/.msf4"]
          detach: true

  database:
      image: postgres:latest
      run:
          env: ["POSTGRES_PASSWORD=msf"]
          publish: ["127.0.0.1:5432:5432"]
          detach: true

  adminer:
      image: clue/adminer
      run:
          link: ["database:postgres"]
          publish: ["127.0.0.1:8080:80"]
          detach: true

  metasploit:
      image: usertaken/metasploit-framework
      run:
          net: host
          tty: true
          interactive: true
          volumes-from: ["loot"]
          cmd: ["/bin/bash"]

``` 
with [Crane](https://github.com/michaelsauter/crane) can be used to spread things out. This is one of the things that need to be addressed for an official Metasploit image.

**Comment by bcook-r7** [comment]
sounds good to me. I had created some database and wrapper helper scripts may also be useful here: https://github.com/bcook-r7/omnibus-metasploit/blob/master/config/patches/metasploit-framework/0001-add-omnibus-helper-scripts.patch

**Comment by bcook-r7** [comment]
I've been thinking about providing an official set of omnibus installer packages as well. A nice thing about the omnibus approach is it bundles the exact version of postgres and everything else you want (ruby, libraries, etc.) into a package and can be fairly distribution-agnostic. But, it's also not mutually exclusive with having a docker image as well. I'm thinking it might be useful if we could publish omnibus metasploit-framework packages for OS X, RPM and DEB, then have a docker image that incorporates the RPM/DEB package along with a minimal base image. 

On a side note, what has your experience been using a docker image with regards to networking with metasploit? Are reverse payload transports might be complicated by the need to configure port forwarding to the container?

**Comment by kernelsmith** [comment]
a vagrant box would be handy too, and would help w/any networking or other stuff that's weird w/Docker's minimal containers.  And/or a packer template, which combined w/an iso (of whatever Linux) one could produce VMware(OVFs or VMX), AMIs, DigOc, Docker, GCE, Parallels, QEMU, VBox (OVF) images, or even all of those.  Packer can also post-process the images, like for compression or stuff like upload it to a cloud provider or vSphere vCenter server or turn it into a vagrant box etc.  Anyways, I have some interest in these devops kinda things if I can help, just let me know

**Comment by UserTaken** [comment]
I noticed a surprisingly amount of open Docker ports while scanning the Internet which anyone can connect to and is the equivalent to root access. I have found up to a dozen vulnerable machines, most being VPS ranging from single core and 512MB RAM to 

```
Containers: 13
Images: 242
Storage Driver: devicemapper
 Pool Name: docker-253:2-4298645046-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 6.493 GB
 Data Space Total: 107.4 GB
 Data Space Available: 100.9 GB
 Metadata Space Used: 11.72 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.136 GB
 Udev Sync Supported: true
 Data loop file: /home/dpeterson/docker_storage/devicemapper/devicemapper/data
 Metadata loop file: /home/dpeterson/docker_storage/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Kernel Version: 3.10.0-229.1.2.el7.x86_64
Operating System: Boss
ANSI_COLOR=
CPUs: 8
Total Memory: 62.73 GiB
Name: 23-25-149-227-static.hfc.comcastbusiness.net
ID: K6H4:AZEN:RR4F:XF4A:BTGQ:ZJQR:MEIT:VFSR:WG26:C7PD:WXLS:SZEK
```

**Comment by nates-r7** [comment]
I just inquired about this internally and was pointed to this thread (thanks @bcook-r7). Kali just released their official docker image (https://registry.hub.docker.com/u/kalilinux/kali-linux-docker/) after you guys had already started this convo. That may provide additional help in building something metasploit specific.
